### PR TITLE
merge aruco changes from 3.4 to 4.x

### PR DESCRIPTION
**Pull request by AleksandrPanov**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/966

Ported PRs:
- https://github.com/opencv/opencv_contrib/pull/3201
- https://github.com/opencv/opencv_contrib/pull/3174

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
> `Merge pull request #XYZ ...`

This message format is not allowed in PRs commits. Remove the first line.
Please take a look on existed port/backport PRs - there are no such things.

Avoid creation of unwanted cross-reference links on PRs (you can check that non-removable trash appears in linked PR).
Commit SHA of original commit is enough.

**Comment by AleksandrPanov** [comment]
> > `Merge pull request #XYZ ...`
> 
> This message format is not allowed in PRs commits. Remove the first line. Please take a look on existed port/backport PRs - there are no such things.
> 
> Avoid creation of unwanted cross-reference links on PRs (you can check that non-removable trash appears in linked PR). Commit SHA of original commit is enough.

fixed

File: modules/aruco/test/test_charucodetection.cpp [hunk 1]
```diff
@@ -681,24 +614,20 @@ TEST(Charuco, testCharucoCornersCollinear_false)
 TEST(Charuco, testBoardSubpixelCoords)
 {
     cv::Size res{500, 500};
-    cv::Mat K = (cv::Mat_<double>(3,3) <<
-        0.5*res.width, 0, 0.5*res.width,
-        0, 0.5*res.height, 0.5*res.height,
-        0, 0, 1);
```
Comment (AleksandrPanov) [inline modules/aruco/test/test_charucodetection.cpp:687]: not useful for this test

**Comment by asmorkalov** [comment]
@AleksandrPanov Is the patch still relevant? Could you close it, if not.

**Comment by alalek** [comment]
@RealOrangeOne what is wrong with PR #3215 ?

**Pull request closed without merging**